ext/standard/array.c

```diff
@@ -3801,29 +3801,37 @@ PHP_FUNCTION(array_change_keys)
         fci.no_separation = 0;
 
         if (str_key) {
-            ZVAL_STR(&params[0], zend_string_copy(str_key));
+            ZVAL_STR_COPY(&params[0], str_key);
         } else {
             ZVAL_LONG(&params[0], num_key);
         }
         ZVAL_COPY(&params[1], value);
 
         if (zend_call_function(&fci, &fci_cache) != SUCCESS || Z_TYPE(result) == IS_UNDEF) {
             zval_dtor(return_value);
-            zval_ptr_dtor(&params[0]);
+            zval_dtor(&params[0]);
+            zval_dtor(&params[1]);
+            efree(params);
             RETURN_NULL();
-        } else {
-            zval_ptr_dtor(&params[0]);
         }
 
+        zval_dtor(&params[0]);
+        zval_dtor(&params[1]);
+
         if (Z_TYPE(result) == IS_STRING) {
-            zend_symtable_update(Z_ARRVAL_P(return_value), Z_STR(result), value);
+            value = zend_symtable_update(Z_ARRVAL_P(return_value), Z_STR(result), value);
+            zval_add_ref(value);
```

Comment:
`zend_symtable_update()` and `zend_hash_index_update()` may return a different pointer than the input value. I gleaned from other examples that we should add a reference to its return value.